paddle/fluid/distributed/collective/ProcessGroupMPI.cc

```diff
@@ -0,0 +1,426 @@
+// Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "paddle/fluid/distributed/collective/ProcessGroupMPI.h"
+#include <chrono>
+#include "paddle/fluid/distributed/collective/Common.h"
+#include "paddle/fluid/platform/cuda_device_guard.h"
+
+constexpr int64_t kWaitBlockTImeout = 10;
+std::map<phi::DataType, MPI_Datatype> mpiDatatype = {
+    {phi::DataType::INT8, MPI_CHAR},
+    {phi::DataType::UINT8, MPI_UNSIGNED_CHAR},
+    {phi::DataType::FLOAT32, MPI_FLOAT},
+    {phi::DataType::FLOAT64, MPI_DOUBLE},
+    {phi::DataType::INT32, MPI_INT},
+    {phi::DataType::INT64, MPI_LONG}};
+
+namespace paddle {
+namespace distributed {
+
+void ProcessGroupMPI::MPITask::FinishMPITaskError(std::exception_ptr eptr) {
+  Finish(eptr);
+}
+
+void ProcessGroupMPI::MPITask::FinishMPITask() { Finish(); }
+
+ProcessGroupMPI::MPIAsyncTask::MPIAsyncTask(
+    MPI_Request request, const std::vector<phi::DenseTensor>& inputs)
+    : ProcessGroup::Task(-1, inputs, CommType::UNKNOWN), request_(request) {
+  memset(&status_, 0, sizeof(status_));
+}
+
+ProcessGroupMPI::MPIAsyncTask::~MPIAsyncTask() {
+  if (request_ != MPI_REQUEST_NULL) {
+    std::cerr << " Task has not completed, try to destruct async mpi task, "
+              << "exit the program." << std::endl;
+    std::terminate();
+  }
+}
+
+bool ProcessGroupMPI::MPIAsyncTask::IsCompleted() {
+  if (request_ == MPI_REQUEST_NULL) {
+    return true;
+  }
+
+  std::unique_lock<std::mutex> lock(pg_global_mutex);
+  int flag = 0;
+  MPI_CHECK(MPI_Test(&request_, &flag, &status_));
+  if (request_ != MPI_REQUEST_NULL) {
+    return false;
+  }
+
+  if (status_.MPI_ERROR != MPI_SUCCESS) {
+    AppearException();
+  }
+
+  return true;
+}
+
+bool ProcessGroupMPI::MPIAsyncTask::Wait(std::chrono::milliseconds timeout) {
+  if (request_ == MPI_REQUEST_NULL) {
+    return true;
+  }
+
+  std::unique_lock<std::mutex> lock(pg_global_mutex);
+  MPI_CHECK(MPI_Wait(&request_, &status_));
+
+  if (status_.MPI_ERROR != MPI_SUCCESS) {
+    AppearException();
+    std::rethrow_exception(exception_);
```

Comment:
return false?

Reply:
DONE

Reply:
What we should return when this function meets error?